### PR TITLE
QA: PHP 7.4 + opcache preloading

### PR DESCRIPTION
**Pull request by michalbundyra**
|    Q          |   A
|-------------- | ------
| QA            | yes

### Description

WIP for now. Based on #49.
Not sure how we can verify if opcache preloading works correctly?

### CODE REVIEW

File: .docker/php/getlaminas.ini [hunk 1]
```diff
@@ -25,3 +25,4 @@ opcache.enable_cli=0
 opcache.save_comments=1
 opcache.load_comments=1
 opcache.error_log=
+opcache.preload = /var/www/preload.php
```
Comment (weierophinney) [inline .docker/php/getlaminas.ini:28]: Considering the docker configuration _is for development_, I don't think we want this.

File: .docker/php/Dockerfile [hunk 2]
```diff
@@ -39,6 +39,7 @@ COPY .docker/php/php-entrypoint /usr/local/bin/php-entrypoint
 # Copy assets
 COPY --from=assets /assets/assets.json /assets/assets.json
 COPY --from=assets /assets/css /assets/css/
+COPY --from=assets /assets/js /assets/js/
```
Comment (weierophinney) [inline .docker/php/Dockerfile:42]: This should likely be done as a separate PR.

File: composer.json [hunk 3]
```diff
@@ -67,6 +75,7 @@
     },
     "scripts": {
         "build": [
+            "./bin/laminas opcache:preload-generate",
```
Comment (weierophinney) [inline composer.json:78]: Definitely not!

You will typically only generate the preload file ONCE, and then make changes to it to customize it to your application. It should NOT be part of the build process.

File: .gitignore [hunk 4]
```diff
@@ -4,6 +4,7 @@ composer.phar
 clover.xml
 coveralls-upload.json
 phpunit.xml
+preload.php
```
Comment (weierophinney) [inline .gitignore:7]: We WANT the preload file checked in. While it has rules to dynamically create the list of files, _it also has an API so you can decide whether or not to ignore files or classes, and/or add more paths_.

**Comment by weierophinney** [comment]
> Not sure how we can verify if opcache preloading works correctly?

You can test it by enabling it. If the application works as expected, it's (a) not breaking anything, and (b) likely working.

From there, benchmarks. Benchmark without preloading enabled, and then with it enabled, and see what happens.

File: preload.php [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+$preloader->paths(
+    'config/',
+    'public/index.php',
+    'src/',
+    'vendor/laminas/',
+    'vendor/mezzio/',
+);
+
+// Preload!
+$preloader->load();
```
Comment (michalbundyra) [inline preload.php:173]: Missing empty line at the end of the file - this is generator issue

/cc @weierophinney

File: composer.json [hunk 6]
```diff
@@ -44,6 +45,7 @@
         "monolog/monolog": "^1.25.3",
         "phly/phly-configfactory": "^1.0.1",
         "phly/phly-event-dispatcher": "^1.0",
+        "phly/phly-opcache-preload": "dev-master",
```
Comment (michalbundyra) [inline composer.json:48]: Should this be then dev dependency?

File: config/config.php [hunk 7]
```diff
@@ -13,6 +13,7 @@
 ];
 
 $aggregator = new ConfigAggregator([
+    \Phly\OpcachePreload\ConfigProvider::class,
```
Comment (michalbundyra) [inline config/config.php:16]: Do we need to have it enabled on prod?

File: composer.json (hunk 6 above)
Comment (weierophinney) [reply composer.json:48]: Honestly, I'm starting to wonder if I should make it a tool to install globally like keep-a-changelog. You'll generally use it once, when starting your project or preparing for initial deployment, and then the `preload.php` file it creates is completely standalone.

You definitely don't need it in production. Development would be fine. Or just install it once, run it, and then remove it.

File: config/config.php (hunk 7 above)
Comment (weierophinney) [reply config/config.php:16]: Nope.

File: preload.php (hunk 5 above)
Comment (weierophinney) [reply preload.php:173]: Send a PR to the phly/phly-opcache-preload repo, please!

**Comment by michalbundyra** [comment]
Probably will be better to move to PHP 8.0 instead.
